### ISSUE TITLE
PosixHeader.prefix and PosixHeader.name are incorrectly typed for POSIX 1003.1-1998 ???

### ISSUE THREAD

**Issue by schnoberts1** [phip1611/tar-no-std#9]
Hi, 

Within PosixHeader we have the fields:
```prefix: StaticCString<155>```

and

```name: StaticSString<FILENAME_MAX_LEN>```

The notes [here](https://www.gnu.org/software/tar/manual/html_node/Standard.html) concur with the above type for ```name```: "The name, linkname, magic, uname, and gname are null-terminated character strings", which lines up with old tar formats which had a max filename length of 99 characters. 

POSIX 1003.1-1988 changed that and allowed up to 256 characters. This requires that ```name``` and ```prefix``` do not have to be null terminated.

To check this I looked at the implementation of gnu tar which has the following logic for handling name/prefix long filenames in read_header in list.c. You can see it uses memcpy not str[n]cpy and explicitly adds a \0 before taking prefixes length as well as adding \0 to ensure the string is terminated post concat of name to the prefix.
```
				if (h->prefix[0] && strcmp(h->magic, TMAGIC) == 0)
				{
					memcpy(np, h->prefix, sizeof h->prefix);
					np[sizeof h->prefix] = '\0';
					np += strlen(np);
					*np++ = '/';
				}
				memcpy(np, h->name, sizeof h->name);
				np[sizeof h->name] = '\0';
```

I think the type of name and prefix should be ArrayString or some optionally null terminated array string if it is to support POSIX 1003.1-1998 headers.

I should explain why these issues are turning up. I am writing a very simple web server for a Raspberry Pico (to learn Rust) and I have embedded the files to be served in the binary as a tar file. rust-no-std is a nice simple library for parsing that file using an ArchiveRef but I'm encountering some compatibility issues as I test the edge cases related to tar file content.

PR: #10

**Comment by phip1611** [comment]
I'm not entirely sure what you are trying to tell me.

So, you mean that fields `prefix` and `name` are allowed to be 156 respectively 100 in size and no terminating null-byte is required? I'm fine with such a change/improvement.

Would you like to work on it?

**Comment by schnoberts1** [comment]
> I'm not entirely sure what you are trying to tell me.

Here's what I observe. The POSIX 1003.1-1098 standard allows for a 256 character filename. The implementation of gnu tar (on reading) allows for both prefix and name to have no null terminator (that is my reading of the code I included above). I can also create such a file within a tar file created by gtar. Here's an example: 

```
(base) andy@ice tar-no-std % tar tf tests/gnu_tar_ustar_long.tar
012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678/
012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678/ABCDEFGHIJABCDEFGHIJABCDEFGHIJABCDEFGHIJABCDEFGHIJABCDEFGHIJABCDEFGHIJABCDEFGHIJABCDEFGHIJABCDEFGHIJ
01234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234/ABCDEFGHIJABCDEFGHIJABCDEFGHIJABCDEFGHIJABCDEFGHIJABCDEFGHIJABCDEFGHIJABCDEFGHIJABCDEFGHIJABCDEFGHIJ
```

and lengths:

```
(base) andy@ice tar-no-std % tar tf tests/gnu_tar_ustar_long.tar | awk '{ print length }' 
100
200
256
```

and the header breakdown is below. Note the "100 of 100", false. Means there are 100 characters in this 100 byte string and it is not null terminated. 

```
PosixHeader { name: ABCDEFGHIJABCDEFGHIJABCDEFGHIJABCDEFGHIJABCDEFGHIJABCDEFGHIJABCDEFGHIJABCDEFGHIJABCDEFGHIJABCDEFGHIJ,100 of 100,false, mode: Mode(Ok(ModeFlags(OwnerRead | OwnerWrite | GroupRead | OthersRead))), uid: 501 [0000765], gid: 20 [0000024], size: 7 [00000000007], mtime: 1693286933 [14473301025], cksum: 19133 [045275 ], typeflag: REGTYPE, linkname: ,0 of 100,true, magic: ustar,5 of 6,true, version: 00,2 of 2,false, uname: andy,4 of 32,true, gname: staff,5 of 32,true, dev_major: cannot parse integer from empty string [], dev_minor: cannot parse integer from empty string [], prefix: 01234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234,155 of 155,false, _pad: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] }
```

> So, you mean that fields `prefix` and `name` are allowed to be 156 respectively 100 in size and no terminating null-byte is required according to the POSIX standard? I'm fine with such a change/improvement.

Yes, that is what I mean. 

> Would you like to work on it?

I am happy to. I have a branch with fixes for this. I have removed the StaticCString and replaced it with a TarFormatString which handles null terminated or fully populated strings. This can be used anywhere in the PosixHeader there is a string. 

While testing this with a tar file created with the Mac OSX tar I noticed the size was corrupted. It turns out there is an edge case in tar files where (for historical reasons) the octal numbers in a tar header can be right padded with a space: 

https://en.wikipedia.org/wiki/Tar_(computing) - _Numeric values are encoded in [octal](https://en.wikipedia.org/wiki/Octal) numbers using ASCII digits, with leading zeroes. For historical reasons, a final NUL or [space](https://en.wikipedia.org/wiki/Space_(punctuation)) character should also be used._

To handle this I replaced ```Size``` with TarFormatOctal that did the handling of the space for completeness.

If that sounds like a reasonable starting point I will create a PR from that branch after a rebase with main. I also took the liberty of creating a Header iterator so you can iterate the PosixHeaders with ``` for h in archive.headers() {....}```. It was useful for debugging. 

Let me know how you'd like me to proceed.

**Comment by phip1611** [comment]
I'm very open for contributions. Personally, I never planned to make this a fully-fledged Tar library. Therefore, functionality and code quality is limited.. (sorry for that). I'm looking forward to your contributions and code cleanups and whatnot :)

**Comment by schnoberts1** [comment]
> ... I never planned to make this a fully-fledged Tar library. 

I understand. The restricted functionality is nice as it means the footprint is small and with directory handling it must cover 90% of use-cases.

**Comment by phip1611** [comment]
Fixed with #10
